include/dlaf/mc/cholesky.h

```diff
@@ -0,0 +1,306 @@
+  // Set up executor on the default queue with default priority.
+  hpx::threads::scheduled_executor matrix_normal_executor =
+      hpx::threads::executors::pool_executor("default", hpx::threads::thread_priority_default);
+
+  auto col_comm_size = grid.colCommunicator().size();
+  auto row_comm_size = grid.rowCommunicator().size();
+
+  // Set up an executor on the mpi pool --> This part need to be added
+  //  hpx::threads::scheduled_executor mpi_executor;
+  //  if (use_pools && m_size > 1) {
+  //    hpx::threads::executors::pool_executor mpi_exec("mpi");
+  //    mpi_executor = mpi_exec;
+  //   }
+  //  else {
+  //    mpi_executor = matrix_HP_executor;
+  //  }
+
+  // Check if matrix is square
+  util_matrix::assert_size_square(mat, "Cholesky", "mat");
+  // Check if block matrix is square
+  util_matrix::assert_blocksize_square(mat, "Cholesky", "mat");
+  // Check compatibility of the communicator grid and the distribution
+  util_matrix::assert_comm_distr(grid, mat, "Cholesky", "mat", "comm");
+
+  const dlaf::matrix::Distribution& distr = mat.distribution();
+
+  SizeType nrtile = mat.nrTiles().cols();
+
+  auto localnrtile_rows = distr.localNrTiles().rows();
+  auto localnrtile_cols = distr.localNrTiles().cols();
+
+  dlaf::common::Pipeline<comm::CommunicatorGrid> serial_comm(std::move(grid));
+
+  // Method only for Lower triangular matrix
+  if (uplo == blas::Uplo::Lower) {
+    for (SizeType k = 0; k < nrtile; ++k) {
+      // Create a placeholder that will store the shared futures representing the panel
+      std::vector<hpx::shared_future<Tile<const T, Device::CPU>>> panel(distr.localNrTiles().rows());
+
+      auto k_rank_row = distr.rankGlobalTile<Coord::Row>(k);
+      auto k_rank_col = distr.rankGlobalTile<Coord::Col>(k);
+
+      if (mat.rankIndex().col() == k_rank_col) {
+        auto k_local_col = distr.localTileFromGlobalTile<Coord::Col>(k);
+
+        hpx::shared_future<Tile<const T, Device::CPU>> kk_tile;
+
+        if (mat.rankIndex().row() == k_rank_row) {
+          auto k_local_row = distr.localTileFromGlobalTile<Coord::Row>(k);
+
+          auto kk = LocalTileIndex{k_local_row, k_local_col};
+
+          // If the diagonal tile is on this node factorize it
+          // Cholesky decomposition on mat(k,k) r/w potrf (lapack operation)
+          hpx::dataflow(matrix_HP_executor, hpx::util::unwrapping(tile::potrf<T, Device::CPU>), uplo,
+                        std::move(mat(kk)));
+
+          // Avoid useless communication if one-column communicator and if on the last column
+          if (col_comm_size > 1 && k != (mat.size().cols() - 1)) {
+            // Broadcast the panel column-wise
+            hpx::dataflow(hpx::util::unwrapping([](auto&& tile, auto&& comm_wrapper) {
+                            dlaf::comm::sync::broadcast::send(comm_wrapper().colCommunicator(), tile);
+                          }),
+                          mat.read(kk), serial_comm());
+          }
+
+          kk_tile = mat.read(kk);
```

Comment:
You can return the tile directly from the lambda.

Reply:
@albestro Why do you want to add an extra dependency on the broadcast?